src/guiguts/widgets.py

```diff
@@ -96,3 +134,32 @@ def grab_focus(
     toplevel.focus_force()
     if focus_widget is not None:
         focus_widget.focus_set()
+
+
+# Dictionary of ToplevelDialog objects, keyed by class name.
+# Used to ensure only one instance of any dialog is created.
+_toplevel_dialogs: dict[str, ToplevelDialog] = {}
+
+TlDlg = TypeVar("TlDlg", bound=ToplevelDialog)
+
+
+def show_toplevel_dialog(
+    dlg_cls: type[TlDlg], root: tk.Tk, title: Optional[str] = None
+) -> TlDlg:
+    """Show the given dialog, or create it if it doesn't exist.
+
+    Args:
+        dlg_cls: Class of dialog to be created - subclass of ToplevelDialog.
+        root: Tk root.
+        title: Dialog title.
+    """
+    global _toplevel_dialogs
```

Comment:
Rather then make this global, make it the python-equiv of a static variable. [The method here](https://stackoverflow.com/a/279597) with the `hasattr` check is the one I use so its entirely self-contained within the function.

Reply:
Thanks @cpeel. That is what I did previously for a similar circumstance, and I tried it again when developing this. I've just had another go after your review. 
The problem is that mypy seems to really struggle with function attributes. You get an error on every line the attribute is used, for example
`src\guiguts\widgets.py:156: error: "Callable[[type[TlDlg], Tk, str | None], TlDlg]" has no attribute "_toplevel_dialogs"  [attr-defined]`
Of course, you can add a suppression comment on each line to get rid of those errors, but it makes it messy. 
Do you think it's OK to keep the global with "_" to flag it as "private", or is there a different technique, or do you think I have misunderstood either the technique or mypy?

Here's the code that gives mypy errors - seven of them:
```
TlDlg = TypeVar("TlDlg", bound=ToplevelDialog)


def show_toplevel_dialog(
    dlg_cls: type[TlDlg], root: tk.Tk, title: Optional[str] = None
) -> TlDlg:
    """Show the given dialog, or create it if it doesn't exist.

    Args:
        dlg_cls: Class of dialog to be created - subclass of ToplevelDialog.
        root: Tk root.
        title: Dialog title.
    """
    if not hasattr(show_toplevel_dialog, "_toplevel_dialogs"):
        show_toplevel_dialog._toplevel_dialogs = {}
    dlg_name = dlg_cls.__name__
    if (
        dlg_name in show_toplevel_dialog._toplevel_dialogs
        and show_toplevel_dialog._toplevel_dialogs[dlg_name].winfo_exists()
    ):
        show_toplevel_dialog._toplevel_dialogs[dlg_name].deiconify()
    else:
        if title is not None:
            show_toplevel_dialog._toplevel_dialogs[dlg_name] = dlg_cls(root, title)  # type: ignore[call-arg]
        else:
            show_toplevel_dialog._toplevel_dialogs[dlg_name] = dlg_cls(root)  # type: ignore[call-arg]
    return show_toplevel_dialog._toplevel_dialogs[dlg_name]  # type: ignore[return-value]
```

Reply:
I looked to see if you could do the equivalent of `# fmt: off` and `# fmt: on` in black to disable/re-enable that specific error for those lines. It doesn't appear possible at the moment - you have to disable the error for the whole module, or you can turn off all checking in a function, neither of which are very satisfactory. [This issue](https://github.com/python/mypy/issues/9483) is requesting the same feature.

Do you think it would be best if I just put the 7 suppression comments in, and not worry about it any more? It's just annoying to have to do that whenever I want a "static", but perhaps that's the price of using mypy?

Reply:
I pushed another commit to show what the code is like with the comments. Happy to go with whichever you think is best.

Reply:
I really hate it when tooling makes you jump through hoops and just make your life harder.

What about if we made the function a class method on TopleveDialog itself? Make `_toplevel_dialogs = {}` a class attribute and then the function becomes something like:
```python
@classmethod
def show_dialog(cls, dlg_cls: type[TlDlg], title: Optional[str] = None) -> TlDlg:
    """Show the given dialog, or create it if it doesn't exist.
    Args:
        dlg_cls: Class of dialog to be created - subclass of ToplevelDialog.
        title: Dialog title.
    """
    # root should be available in a higher-level parent, right?
    dlg_name = dlg_cls.__name__
    if (
        dlg_name in _toplevel_dialogs
        and _toplevel_dialogs[dlg_name].winfo_exists()
    ):
        _toplevel_dialogs[dlg_name].deiconify()
    else:
        if title is not None:
            _toplevel_dialogs[dlg_name] = dlg_cls(root, title)  # type: ignore[call-arg]
        else:
            _toplevel_dialogs[dlg_name] = dlg_cls(root)  # type: ignore[call-arg]
    return _toplevel_dialogs[dlg_name]  # type: ignore[return-value]
```

Which becomes effectively a factory method.

Reply:
Thanks - I've had a bit of a look at that, and can have a closer look tomorrow.
I also experimented with keeping it as a function and function attribute, but using try/except, which resulted in fewer uses of the attribute, so fewer suppression comments (though I could probably do similar with the `hasattr` version by using a local `dlg` variable like I did in this version). 

Please could you let me know which version you prefer: the hasattr (maybe with local variable), the class method, or the try/except version below. I have a slight inclination towards the hasattr version, but would be OK with any of them tbh. Really appreciate all the advice, and learning different ways to do things - thanks.

```python
def show_toplevel_dialog(
    dlg_cls: type[TlDlg], root: tk.Tk, title: Optional[str] = None
) -> TlDlg:
    """Show the given dialog, or create it if it doesn't exist.

    Args:
        dlg_cls: Class of dialog to be created - subclass of ToplevelDialog.
        root: Tk root.
        title: Dialog title.

    Attributes:
        _toplevel_dialogs: Dictionary of ToplevelDialog objects, keyed by class name.
            Used to ensure only one instance of any dialog is created.
    """
    dlg_name = dlg_cls.__name__
    try:
        dlg = show_toplevel_dialog._toplevel_dialogs[dlg_name]  # type: ignore[attr-defined]
    except (AttributeError, KeyError) as exc:
        # Dialog not found in dict
        if isinstance(exc, AttributeError):
            # First time - create empty dict
            show_toplevel_dialog._toplevel_dialogs = {}  # type: ignore[attr-defined]
        dlg = None
    if dlg and dlg.winfo_exists():
        # Dialog already exists, so show it
        dlg.deiconify()
    else:
        # Dialog doesn't currently exist, so create it
        dlg = dlg_cls(root) if title is None else dlg_cls(root, title)  # type: ignore[call-arg]
        show_toplevel_dialog._toplevel_dialogs[dlg_name] = dlg  # type: ignore[attr-defined]
    return dlg
```

Reply:
> Please could you let me know which version you prefer: the hasattr (maybe with local variable), the class method, or the try/except version below. I have a slight inclination towards the hasattr version

The `hasattr` version is great with me and if that's the one you prefer I say go for it! I would prefer a non-global method but am largely ambivalent about the other options.

Reply:
@cpeel - many thanks. After looking in more detail at your class method solution, I think that makes most sense. In [these docs](https://docs.python.org/3/howto/descriptor.html#class-methods), it says "One use for class methods is to create alternate class constructors. For example, the classmethod [dict.fromkeys()](https://docs.python.org/3/library/stdtypes.html#dict.fromkeys) creates a new dictionary from a list of keys."
That seems similar to what we're doing here.
So, I've pushed a commit more or less with your code plus a couple of type declaration edits.

Regarding "root", TIL you don't need to give root as an argument when creating a Toplevel widget, which means I've been able to remove use of it from several places. So, I've pushed another commit for that. Thanks for prompting me to investigate that.

Hopefully ready for a final check of the code by you now, thanks.

Also, @srjfoo, if you could do one more quick check that Search is basically working, especially that you can't get two search dialogs up at once, that would be great, thank you.

Reply:
Oh yeah, the class method looks super clean. 👍